### PR TITLE
Attempts to add Type hints to Clock and Event.py

### PR DESCRIPTION
**Pull request by gran4**
Can the library typing_extensions be added to pyglet?

### CODE REVIEW

**Comment by pushfoo** [comment]
> Can the library typing_extensions be added to pyglet?

tl;dr I've also asked about it and been told no. It's annoying, but there are good reasons for it.

Banning mandatory external dependencies improves portability and eliminates a lot of potential problems, including upstream license or requirement changes breaking the project. For example, arcade had to drop 3.7 support early because a crucial dependency forced us to.

I asked about using typing_extensions [last year](https://discord.com/channels/438622377094414346/477394331112701952/1006813955685089370):

> pushfoo — 08/10/2022 2:39 AM
> I think -> Vec2 technically works for mypy 
> I wish I didn't check the PEP listing though
> knowing that the Self annotation exists makes me want to use the typing backports > module 
> https://github.com/python/typing_extensions
> Benjamin — 08/10/2022 2:41 AM
> unfortunately, external dependencies are a deal-breaker for pyglet.

[Additional details](https://discord.com/channels/438622377094414346/477394331112701952/1006815338446475294):
> Benjamin — 08/10/2022 2:44 AM
pyglet actually doesn't have any dependencies - only optional modules that it will use if they're available.

**Comment by gran4** [comment]
> tl;dr I've also asked about it and been told no. It's annoying, but there are good reasons for it.

What should I do instead?

**Review by pushfoo — CHANGES_REQUESTED** [review]
> What should I do instead?

tl;dr Slow down and give yourself time to read & think. Also install mypy or pyright locally and use them before pushing since pyglet's CI doesn't run type checking yet.

The rest of the comment describes what I did to write the comments on your code changes.

### 1 / 3 : Take your time to read & experiment

Your recent PRs are more focused & concise, which is good.

However, you've started playing on hard mode by working on some of the oldest & most crucial library code in pyglet (2006). The code's age isn't a bad thing! The design works!

Trying to modernize this sort of codebase means you're taking on the responsibility of figuring out what's going on. This means:

1. Reading the code and experimenting with it
2. Reading any relevant doc, such as Python & library documentation
3. Reading supporting theory for things like shaders

For example, I usually have Python and library doc open at all times when I'm working. This includes while I was reviewing this PR. Sometimes, it's also necessary to search through old commits, bug reports, or PRs to understand what's wrong.

### 2 / 3 : Run type checking & unit tests before committing

After installing pyglet locally in dev mode, you can run unit tests locally with the following:

```console
pytest tests/unit
```
This will catch the sort of errors you've been seeing in the CI run, but without important benefits:
1. No waiting for the GitHub runner to become available & set up the environment.
2. If you're using an IDE like VSCode or PyCharm, file paths in exception traceback will be clickable links you can use to jump to the location of the error.

You should also set up linting and type checking locally because pyglet may not run thorough checking.
```console
pip install mypy
```

Afterwards, run it against specific files like this:
```console
mypy path/to/changed/file.py
```
I think ruff and pyright work the same way, but are much stricter by default.

### 3 / 3 : Do the research to answer your own questions when possible

I'm serious about this. Keep doc open in one window, code in the other. Search through  the GitHub issues of projects you use to look for relevant tickets, including closed ones.

A second monitor really helps if you can get one, even a cheap old 720p panel only big enough to fit a terminal.

File: pyglet/event.py [hunk 1]
```diff
@@ -158,14 +165,15 @@ def register_event_type(cls, name):
         cls.event_types.append(name)
         return name
 
-    def push_handlers(self, *args, **kwargs):
-        """Push a level onto the top of the handler stack, then attach zero or
+    def push_handlers(self, *args: Tuple, **kwargs: Optional[object, Callable]) -> None:
+        """
+        Push a level onto the top of the handler stack, then attach zero or
```
[the remark this answers is not in the document]
Comment (gran4) [reply pyglet/event.py:170]: Yeah

File: pyglet/clock.py [hunk 2]
```diff
@@ -342,17 +349,41 @@ def _get_nearest_ts(self):
...
+    def _get_soft_next_ts(self, last_ts: float, interval: float) -> float:
+        """Attempts to return evenly scheduled functions
+
+        :Parameters:
+            `last_ts` : float
+                The last time stamp
+            `interval` : float
+                Look at the information about scheduling, it is down 
+
+        :rtype: float
```
[the remark this answers is not in the document]
Comment (gran4) [reply pyglet/clock.py:372]: @pushfoo  It seems to be the same.

File: pyglet/event.py [hunk 3]
```diff
@@ -401,15 +409,29 @@ def dispatch_event(self, event_type, *args):
         return False
 
     @staticmethod
-    def _raise_dispatch_exception(event_type, args, handler, exception):
-        # A common problem in applications is having the wrong number of
```
Comment (benmoran56) [inline pyglet/event.py:405]: This should be left as a block comment. It's an internal method that is not intended to be seen by users. The block comment contains notes for future programmers, not intended for user documentation.

**Review by benmoran56 — CHANGES_REQUESTED** [review]

File: pyglet/clock.py [hunk 4]
```diff
@@ -63,37 +63,44 @@ def move(dt, velocity, sprite):
...
+    def __init__(self, func: Callable, interval: float, last_ts: float, next_ts: float, args: Any, kwargs: Any) -> None:
         self.func = func
         self.interval = interval
         self.last_ts = last_ts
         self.next_ts = next_ts
         self.args = args
         self.kwargs = kwargs
 
-    def __lt__(self, other):
+    def __lt__(self, other: Union[_ScheduledIntervalItem, float]) -> bool:
```
[the remark this answers is not in the document]
Comment (gran4) [reply pyglet/clock.py:103]: look at line 106 `return self.next_ts < other`. other is a float

[the remark this answers is not in the document]
Comment (gran4) [reply pyglet/clock.py:103]: @benmoran56 I want you to see this if you have not already.
